terminals/Altapay_Test_Terminal.class.php

```diff
@@ -10,401 +10,509 @@
+				$order->payment_complete();
+				update_post_meta( $order_id, '_transaction_id', $txnId );
+				update_post_meta( $order_id, '_cardno', $cardNo );
+				update_post_meta( $order_id, '_credit_card_token', $credToken );
+				update_post_meta( $order_id, '_credit_card_brand', $creditCardCardBrand );
+				update_post_meta( $order_id, '_credit_card_expiry_date', $creditCardExpiryDate );
+			}
+
+			// Redirect to Order Confirmation Page
+			if ( $type === 'paymentAndCapture' && $requireCapture === 'true' ) {
+				$api = $this->altapayApiLogin();
+				if ( ! $api || is_wp_error( $api ) ) {
+					echo '<p><b>' . __( 'Could not connect to AltaPay!', 'altapay' ) . '</b></p>';
+					return;
+				}
+				$api = new CaptureReservation( $this->getAuth() );
+				$api->setAmount( (float) number_format( $amount, 2, '.', '' ) );
+				$api->setTransaction( $txnId );
+
+				/** @var CaptureReservationResponse $response */
+				try {
+					$response = $api->call();
+				} catch ( ResponseHeaderException $e ) {
+					error_log( 'Response header exception ' . $e->getMessage() );
+				} catch ( \Exception $e ) {
+					error_log( 'Response header exception ' . $e->getMessage() );
+				}
+			}
+			$redirect = $this->get_return_url( $order );
+			wp_redirect( $redirect );
+			exit;
+		}
+	}
+
+	/**
+	 * @param array   $addressInfo
+	 * @param Address $address
+	 *
+	 * @return object
+	 */
+	private function createAddressObject( $addressInfo, $address ) {
+		$address->Firstname  = $addressInfo['firstname'];
+		$address->Lastname   = $addressInfo['lastname'];
+		$address->Address    = $addressInfo['address'];
+		$address->City       = $addressInfo['city'];
+		$address->PostalCode = $addressInfo['postcode'];
+		$address->Region     = $addressInfo['region'] ?: '0';
+		$address->Country    = $addressInfo['country'];
+
+		return $address;
+	}
+
+	/**
+	 * @param WC_Order $order
+	 *
+	 * @return Customer
+	 * @throws Exception
+	 */
+	public function setCustomer( $order ) {
+		$address        = new Address();
+		$altapayHelpers = new Helpers\AltapayHelpers();
+		$billingInfo    = array(
+			'firstname' => $order->get_billing_first_name(),
+			'lastname'  => $order->get_billing_last_name(),
+			'address'   => $order->get_billing_address_1(),
+			'postcode'  => $order->get_billing_postcode(),
+			'city'      => $order->get_billing_city(),
+			'region'    => $order->get_billing_state(),
+			'country'   => $order->get_billing_country(),
+		);
+		$shippingInfo   = array(
+			'firstname' => $order->get_shipping_first_name(),
+			'lastname'  => $order->get_shipping_last_name(),
+			'address'   => $order->get_shipping_address_1(),
+			'postcode'  => $order->get_shipping_postcode(),
+			'city'      => $order->get_shipping_city(),
+			'region'    => $order->get_shipping_state(),
+			'country'   => $order->get_shipping_country(),
+		);
+		if ( $order->get_billing_address_1() ) {
+			$address = $this->createAddressObject( $billingInfo, $address );
```

Comment:
You already have the address object in that variable.
```suggestion
			$this->createAddressObject( $billingInfo, $address );
```